Add WEB_FEATURES_MANIFEST.json generation

jcscottiii opened the pull request:
# About

Creates a new command: `./wpt web-features-manifest`

This command will generate an artifact that maps web features to web platform tests

The artifact created will be named WEB_FEATURES_MANIFEST.json.

The format of the file is a simple dictionary.
- The key is the web feature key from https://github.com/web-platform-dx/web-features/tree/main/feature-group-definitions
- The value is a list of test urls that apply to the web feature key. We use the urls (instead of path) because wpt.fyi can use those to easily match the test

This PR adds logic to traverse the file system and build that dictionary.

# Other changes
- tools/manifest/tests/test_sourcefile.py - Add META.yml and WEB_FEATURES.yml to be excluded as test files

# Note to reviewers
- There will be a future PR that adds logic in tools/ci/manifest_build.py to generate this file alongside the existing MANIFEST.json. I did a partial upload of the assets created in https://github.com/jcscottiii/wpt/releases
  - There will be a new process in wpt.fyi that uses that file. Draft PR: https://github.com/web-platform-tests/wpt.fyi/pull/3645

jcscottiii committed:
Add WEB_FEATURES_MANIFEST.json generation

This will allow mapping of web features to web platform tests

Creates a new command: ./wpt web-features-manifest

It will create a file called WEB_FEATURES_MANIFEST.json inside of the
repo_root.

The format of the file is a simple dictionary.

The key is the web feature key from https://github.com/web-platform-dx/web-features/tree/main/feature-group-definitions

The value is a list of test paths that apply to the web feature key

fix relative import

add path argument

improve manifest file

Update form of the web_features manifest file

wpt.fyi needs the url instead of path for test comparison

simplify the manifest file

change 'web features' to 'web-features'

resolve https://github.com/web-platform-tests/wpt/pull/43828#discussion_r1455526145

resolve https://github.com/web-platform-tests/wpt/pull/43828#discussion_r1455543166

resolve https://github.com/web-platform-tests/wpt/pull/43828#discussion_r1455461870

resolve https://github.com/web-platform-tests/wpt/pull/43828#discussion_r1455519572

resolve https://github.com/web-platform-tests/wpt/pull/43828/files/8f520a2d9e815973e53598d582ed455391be98f1#r1455453034

validate schema

use os.listdir instead of os.walk

Use mypy ignore-errors

Similar to tools/ci/tc/tests/test_valid.py that also uses jsonschema

jcscottiii commented on tools/manifest/sourcefile.py:
```diff
@@ -29,6 +29,10 @@
 
+# Cannot do `from ..metadata.webfeatures.schema import WEB_FEATURES_YML_FILENAME`
+# because relative import beyond toplevel throws *ImportError*!
+from metadata.webfeatures.schema import WEB_FEATURES_YML_FILENAME  # type: ignore
```
Note for reviewers: This is similar to this section:

https://github.com/web-platform-tests/wpt/blob/7c6e57968932c49879182ae0fbbe1a949484ccee/tools/manifest/vcs.py#L12-L14

foolip commented on tools/web_features/manifest.py:
```diff
@@ -0,0 +1,249 @@
+        cmd_cfg (CmdConfig): The configuration for the command-line options.
+        rel_dir_path (str): The relative path of the directory to process.
+        result (WebFeaturesMap): The object to store the mapping results.
+        visited_dirs (Set[str], optional): A set of directories that have already been processed. Defaults to set().
```
How can we end up visiting the same directory twice? The recursive call is with a subdirectory, and there are no cycles, right?

jcscottiii replied:
Thanks for bringing this up.

This was my first time really using os.walk. During my experiments I saw that each iteration of os.walk will traverse the filesystem from top to bottom already [1]. And since I'm also calling my function again on the subdirs, the same folder would eventually be called on the next iteration of os.walk.


[1] `for current_dir, sub_dirs, _ in os.walk(os.path.join(cmd_cfg.repo_root, rel_dir_path)):`

[2] 

```
        for sub_dir in sub_dirs:
            map_tests_to_web_features(
                cmd_cfg,
                os.path.join(rel_dir_path, sub_dir),
                result,
                visited_dirs,
                inherited_features
            )
```
As a result, I needed that `visited_dirs` variable.

But I see what you are saying. I'm going to remove the use of os.walk and only do listdir (or something similar). That way I won't need that visited_dirs variable.

foolip commented on tools/web_features/manifest.py:
```diff
@@ -0,0 +1,249 @@
+                {
+                    "version": 1,
+                    "data": feature_map,
+                    "config": {"url_base": cmd_cfg.url_base}
```
Is `cmd_cfg.url_base` always a relative path, or can we end up including a full path here? And do we really need "url_base", can it be used for anything once the manifest has been sent to another system like wpt.fyi?

jcscottiii replied:
You're right. We can remove this. It was not used in my implementation in wpt.fyi

foolip commented on tools/web_features/web_feature_map.py:
```diff
@@ -0,0 +1,136 @@
+
```
Extra blank line

foolip commented on tools/web_features/manifest.py:
```diff
@@ -0,0 +1,249 @@
+         "type":"object",
+         "additionalProperties":false,
+         "properties":{
+            "url":{
```
This is a (relative) URL, but I wonder if "path" would work here? It seems like the more common terminology I've heard people use when referring to these strings, and matches wpt.fyi's `path:/.../` query syntax, for example.

jcscottiii replied:
I agree that `path` would be better. However, the variable name comes from the existing `url` field in URLManifestItem [1] [2]. As a result, I thought it would be better to keep it consistent. Let me know your thoughts.

[1] The original value:
- https://github.com/web-platform-tests/wpt/blob/d4d9e95b565eadcd672aeab45b70dbd924020f6c/tools/manifest/item.py#L112-L117

[2] Where we copy the value
- https://github.com/web-platform-tests/wpt/pull/43828/files#diff-3a95149a0e08f798d0a1b597ed2d0b9e4477f82192fb30addacdc46b958daf52R16-R18

foolip replied:
Both would be OK, but I still think we should go with `path`, in particular because of the wpt.fyi search syntax.

foolip commented on tools/web_features/web_feature_map.py:
```diff
@@ -0,0 +1,136 @@
+    def to_dict(self) -> Dict[str, List[Dict[str, str]]]:
+        """
+        Returns:
+            The plain python dictionary representation of the map.
```
```suggestion
            The plain dictionary representation of the map.
```

foolip commented on tools/web_features/web_feature_map.py:
```diff
@@ -0,0 +1,136 @@
+
+class WebFeaturesMap:
+    """
+    Stores a mapping of web features to their associated test files.
```
Judging by the unit tests it's a map to test paths, not file paths, right? Or what happens with `.any.js` tests?

jcscottiii commented on tools/web_features/manifest.py:
```diff
@@ -0,0 +1,249 @@
+    parser = argparse.ArgumentParser(
+        description="Maps tests to web features within a repo root."
+    )
+    parser.add_argument("--repo-root", type=str,
```
We can probably simplify this.

jcscottiii commented on tools/web_features/manifest.py:
```diff
@@ -0,0 +1,249 @@
+throughout the wpt source code. For more details about the file, check out RFC 163
+https://github.com/web-platform-tests/rfcs/pull/163.
+
+The file written is a JSON file. The format is
```
I will create a new file called MANIFEST_SCHEMA.json (so that it won't be counted as a test file) https://github.com/web-platform-tests/wpt/blob/e7d1fb45fdf2ac5d5126e87cc5ebec7ef6c6b131/tools/manifest/sourcefile.py#L303

jcscottiii replied:
Done

foolip commented on tools/web_features/web_feature_map.py:
```diff
@@ -0,0 +1,135 @@
+        Adds a web feature and its associated test paths to the map.
+
+        Args:
+            feature: The name of the web feature.
```
```suggestion
            feature: The web-features identifier.
```

foolip reviewed (approved):
LGTM with `path` instead of `url` and the minor suggestions I had.

foolip commented:
So I finally tried this locally to see what the actual JSON looks like. Here it is, trimmed for clarity:

```json
{
    "version": 1,
    "data": {
        "idle-detection": [
            {
                "url": "/idle-detection/basics.tentative.https.window.html"
            },
            {
                "url": "/idle-detection/idle-detection-allowed-by-permissions-policy-attribute-redirect-on-load.https.sub.html"
            },
            {
                "url": "..."
            }
        ],
        "webtransport": [
            {
                "url": "/webtransport/back-forward-cache-with-closed-webtransport-connection-ccns.https.tentative.window.html"
            },
            {
                "url": "/webtransport/back-forward-cache-with-closed-webtransport-connection.https.window.html"
            },
            {
                "url": "..."
            },
        ],
        "async-clipboard": [
            {
                "url": "/clipboard-apis/async-custom-formats-write-fail.tentative.https.html"
            },
            {
                "url": "/clipboard-apis/async-custom-formats-write-read-web-prefix.tentative.https.html"
            },
            {
                "url": "..."
            }
        ]
    }
}
```

I think we should make two changes here:

- Ensure feature identifiers are sorted, so that the output is more stable
- Drop the object wrapping and just have an array of strings instead

It would look like this:

```json
{
    "version": 1,
    "data": {
        "async-clipboard": [
            "/clipboard-apis/async-custom-formats-write-fail.tentative.https.html",
            "/clipboard-apis/async-custom-formats-write-read-web-prefix.tentative.https.html",
            "..."
        ],
        "idle-detection": [
            "/idle-detection/basics.tentative.https.window.html",
            "/idle-detection/idle-detection-allowed-by-permissions-policy-attribute-redirect-on-load.https.sub.html",
            "..."
        ],
        "webtransport": [
            "/webtransport/back-forward-cache-with-closed-webtransport-connection-ccns.https.tentative.window.html",
            "/webtransport/back-forward-cache-with-closed-webtransport-connection.https.window.html",
            "..."
        ]
    }
}
```

@jcscottiii WDYT?

jcscottiii committed:
Apply suggestions from code review

Co-authored-by: Philip Jägenstedt <philip@foolip.org>

jcscottiii committed:
resolve https://github.com/web-platform-tests/wpt/pull/43828#discussion_r1458649867

jcscottiii committed:
Resolve https://github.com/web-platform-tests/wpt/pull/43828#issuecomment-1900071334

- Ensure feature identifiers are sorted, so that the output is more stable
- Drop the object wrapping and just have an array of strings instead
- Adjust the schema file and tests to reflect that

jcscottiii commented on tools/web_features/manifest.py:
```diff
@@ -0,0 +1,195 @@
+paths  from the test files, and writes them to a manifest file. The manifest
+file maps web-feature identifiers to their corresponding test paths.
+
+The file written is a JSON file. An example file looks like:
```
Whoops! Resolved in https://github.com/web-platform-tests/wpt/pull/43828/commits/aa95e172783d76bfdc31ac4b10f241797f0b9d0e

Thanks!

jcscottiii commented:
I like those changes proposed in https://github.com/web-platform-tests/wpt/pull/43828#issuecomment-1900071334

I have pushed code to incorporate it. Upon clean CI build, I will merge.
